Add action for configuring default action of EditableText.onTapUpOutside

Hannnes1 opened the pull request:
This PR adds an `Action` for configuring a default action of `EditableText.onTapUpOutside`. This is the equivalent to what https://github.com/flutter/flutter/pull/150125 did for `EditableText.onTapOutside`.

Fixes https://github.com/flutter/flutter/issues/162212

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

Hannnes1 committed:
Add action for onTapUpOutside

Hannnes1 committed:
Add test

Hannnes1 committed:
Add example

Hannnes1 committed:
Fix formatting

justinmc commented on packages/flutter/lib/src/widgets/text_editing_intents.dart:
```diff
@@ -407,3 +407,38 @@ class EditableTextTapOutsideIntent extends Intent {
+/// {@tool dartpad}
+/// A common requirement is to unfocus text fields when the user taps outside of
+/// it. For UX reasons, it's often desirable to only unfocus when the user taps
+/// outside of the text field, but not when they scroll.
+/// To achieve this, you can override the default behavior of
```
Nit: Empty line `///` between these two paragraphs.

justinmc commented on packages/flutter/test/widgets/editable_text_test.dart:
```diff
@@ -11969,6 +11969,45 @@ void main() {
+    );
+    await tester.pump();
+    await tester.tap(find.byKey(key), warnIfMissed: false);
+    await tester.pumpAndSettle();
```
Nit: If this works with `pump` instead of `pumpAndSettle`, I would switch to that in order to have less uncertainty in the test.

Hannnes1 replied:
I copied this from the onTapOutside test (the test just above this one), but when running it locally I'm not sure if the pump there is needed at all.

justinmc commented on examples/api/lib/widgets/text_editing_intents/editable_text_tap_up_outside_intent.0.dart:
This is a good, practical example, thanks for adding it!

Hannnes1 replied:
Thanks you!

justinmc reviewed (approved):
LGTM with nits :+1: . Thanks for jumping on this PR.

Hannnes1 committed:
Add line break to comment

Hannnes1 committed:
Minor test change

Hannnes1 committed:
Remove duplicate MaterialApp from example

Hannnes1 committed:
Fix failing test on web

This test is copied from the test for onTapOutside. That test changes
the unfocusing behaviour, and therefore verifies that.
I had erroneously left that check in my test, even if it doesn't change
the default unfocusing behaviour.

justinmc commented on examples/api/lib/widgets/text_editing_intents/editable_text_tap_up_outside_intent.0.dart:
```diff
@@ -0,0 +1,117 @@
+              onInvoke: _handlePointerUp,
+            ),
+          },
+          child: ListView(
+            children: <Widget>[
+              TextField(focusNode: FocusNode()),
+              ...List<Widget>.generate(50, (int index) => Text('Item $index')),
```
Nit: This might be better as a ListView.builder since you are using List.generate? But either way is fine.

justinmc commented on packages/flutter/test/widgets/editable_text_test.dart:
```diff
@@ -11969,6 +11969,43 @@ void main() {
+        ),
+      ),
+    );
+    await tester.pump();
```
Nit: Is this pump necessary?

justinmc reviewed (approved):
Renewing my LGTM :+1: . I'll try to find a second reviewer.

victorsanni commented on examples/api/lib/widgets/text_editing_intents/editable_text_tap_up_outside_intent.0.dart:
```diff
@@ -0,0 +1,117 @@
+    // the pointer down and pointer up events later.
+    latestPointerDownEvent = intent.pointerDownEvent;
+
+    // Match de default behavior of unfocusing on tap down on desktop platforms
```
```suggestion
    // Match the default behavior of unfocusing on tap down on desktop platforms
```

victorsanni commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -1563,6 +1563,11 @@ class EditableText extends StatefulWidget {
   /// group when the text field is focused.
   ///
+  /// If this is null, [EditableTextTapUpOutsideIntent] will be invoked. In the
+  /// default implementation, this is a no-op. To change this behavior, a
+  /// callback may be set here or [EditableTextTapUpOutsideIntent] may be
+  /// overridden.
```
Really small documentation nit: It might be easier to read this in active vs. passive voice. So instead:

```suggestion
  /// default implementation, this is a no-op. To change this behavior, set a
  /// callback here or override [EditableTextTapUpOutsideIntent].
```

or something similar.

victorsanni commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -1573,6 +1578,8 @@ class EditableText extends StatefulWidget {
   ///
   ///  * [TapRegion] for how the region group is determined.
   ///  * [onTapOutside], which is called for each tap down.
+  ///  * [EditableTextTapOutsideIntent] for the intent that is invoked if
```
Really small nit: change to active to make it easier to read. So instead:

```suggestion
  ///  * [EditableTextTapOutsideIntent], the intent that is invoked if
```

victorsanni reviewed (approved):
Thanks for the PR. I left some small documentation nits.

Hannnes1 committed:
Apply suggestions from code review

Co-authored-by: Victor Sanni <victorsanniay@gmail.com>

justinmc commented:
@Hannnes1 Heads up there is an analyzer failure.

Hannnes1 commented:
Thanks @justinmc. Seems like it was a flaky test or something. Rebased master again, and everything passes now.
